### PR TITLE
Speckle2Self implementation

### PR DESCRIPTION
**Pull request by tristan-deep**
zea implementation of the Speckle2Self algorithm

With some regressions test it seems this implementation is the same as the original, however should check if preprocessing steps etc. are being done correctly. Outputs seem consistent with paper results. See [example notebook](https://zea--377.org.readthedocs.build/en/377/notebooks/models/speckle2self_despeckling_example.html).

[arXiv](https://arxiv.org/abs/2507.06828) &nbsp; [GitHub](https://github.com/noseefood/speckle2self)

### Short axis (in-house)
speckle2self_output

### Long axis (in-house)
speckle2self_output

### Short axis (picmus)
speckle2self_output

Model works quite well on their own test data. Should try to mimick preprocessing steps to match. This does seem to be B-Mode data, unlike their statement: _For optimal performance, please use ultrasound envelope data (not B-mode) with a minimum resolution of 512×512._

speckle2self_output_test

```python
import numpy as np
import keras
import matplotlib.pyplot as plt

from zea.models.speckle2self import Speckle2Self

model = Speckle2Self.from_preset("hf://zeahub/speckle2self-invivo")

data = np.load("/workspace/temp/speckle2self/test_data.npy")

despeckled = model(data[0, None, ..., None])
despeckled = keras.ops.convert_to_numpy(despeckled)
despeckled = despeckled.squeeze()

fig, axes = plt.subplots(1, 2, figsize=(8, 4), squeeze=False)
axes[0, 0].imshow(data[0], cmap="gray")
axes[0, 0].set_title("Input")
axes[0, 1].imshow(despeckled, cmap="gray")
axes[0, 1].set_title("Despeckled")
plt.tight_layout()
plt.savefig("speckle2self_output_test.png", bbox_inches="tight", dpi=100)

```

## Summary by CodeRabbit

* **New Features**
  * Added the Speckle2Self ultrasound despeckling model to the models API with presets, multiple weight backends, and an optional inference fallback.

* **Documentation**
  * Added a GPU/TPU tutorial demonstrating the end-to-end despeckling workflow, visualization, and example output.
  * Clarified notebook wording and updated a citation to “et al.”; added guidance for integrating custom models.

* **Tests**
  * Added tests covering Speckle2Self input/output shape behavior.

### CODE REVIEW

**Comment by tristan-deep** [comment]
Hi @noseefood! Attempted to integrate Speckle2Self ([GitHub](https://github.com/noseefood/speckle2self)) into [zea](https://joss.theoj.org/papers/10.21105/joss.09881). See [example notebook](https://zea--377.org.readthedocs.build/en/377/notebooks/models/speckle2self_despeckling_example.html) on some of our data. Could you comment on whether the pre/post processing steps for the data are correctly implemented for this method? And is the output as expected or can we improve?

File: zea/models/speckle2self.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+                import onnxruntime
+            except ImportError as e:
+                raise ImportError(
+                    "Install onnxruntime or provide a .weights.h5 file for Speckle2Self"
+                ) from e
+            filename = loader.get_file("model.onnx")
+            self._onnx_sess = onnxruntime.InferenceSession(filename)
+        elif backend == "torch":  # pragma: no cover
+            filename = loader.get_file("model.pth")
+            self._load_from_pth(filename)
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply zea/models/speckle2self.py:578]: Fixed

[the remark this answers is not in the document]
Comment (tristan-deep) [reply zea/models/speckle2self.py:578]: Fixed

File: zea/models/speckle2self.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+        """
+        assert inputs.ndim == 4, (
+            f"Input should have 4 dimensions (B, H, W, C), but has {inputs.ndim}."
+        )
+
+        assert inputs.shape[-1] == 1, f"Input should have 1 channel, but has {inputs.shape[-1]}."
+
+        original_size = ops.shape(inputs)[1:3]
+        inputs = ops.image.resize(inputs, [INFERENCE_SIZE, INFERENCE_SIZE])
+        inputs = zea.func.normalize(inputs, output_range=(0, 1))
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply zea/models/speckle2self.py:493]: Could be, for now i only got good results at this resolution so fixing it for that.

File: zea/models/__init__.py [hunk 3]
```diff
@@ -82,6 +83,52 @@
 4. Add a preset for the model in :mod:`zea.models.presets`. This basically allows you to have multiple weights presets for a given model architecture.
 5. Make sure to register the presets in your model module by importing the presets module and calling ``register_presets`` with the model class as an argument.
 6. Lastly, add the model to the :mod:`zea.models` package by importing it in the :file:`__init__.py` file.
+
+Adding non-Keras (custom) models
```
Comment (OisinNolan) [inline zea/models/__init__.py:87]: Nice addition to the docs!

File: zea/models/speckle2self.py [hunk 4]
```diff
@@ -0,0 +1,622 @@
+"""Speckle2Self self-supervised speckle reduction model for ultrasound images.
+
+Implements the SpeckleReductionNet architecture from the Speckle2Self paper
+(Romaguera et al., arXiv:2507.06828) as a native Keras 3 model.
```
Comment (OisinNolan) [inline zea/models/speckle2self.py]: Maybe this would be better as a link?

Image

Comment (OisinNolan) [reply zea/models/speckle2self.py]: Also it looks like 'Romaguera et al.' is hallucinated? The authors are 'Xuesong Lia,b, Nassir Navaba,b, Zhongliang Jianga,b,∗'

Comment (OisinNolan) [reply zea/models/speckle2self.py]: Also don't see the phrase 'SpeckleReductionNet' in the paper -- is this from some other paper?

**Review by OisinNolan — COMMENTED** [review]
Very nice addition! Just spotted some potential hallucinations in the docs. Also I was thinking maybe a cross-sectional carotid view could be a nice example for the notebook once we can open source the whole carotid dataset, since it matches their data better.

File: zea/models/speckle2self.py (hunk 4 above)
Comment (tristan-deep) [reply zea/models/speckle2self.py]: Woops this should be correct, I copied the layout from some other rst page.

Comment (tristan-deep) [reply zea/models/speckle2self.py]: SpeckleReductionNet is mentioned in their codebase [here](https://github.com/noseefood/speckle2self/blob/dd09fb3195c6a92665c93cbdefb9cec22e66d8f5/networks/srn/net.py#L210)!

**Comment by tristan-deep** [comment]
Thanks @OisinNolan ! I added the short axis view result as well to the PR description (`6_cross_2cm_R_0000.hdf5`), will update the notebook after we release that!

**Review by OisinNolan — APPROVED** [review]

**Comment by noseefood** [comment]
Hi, zea teams. Thanks a lot for integrating Speckle2Self into zea! I’m very happy to see this implementation. Last week I was not in the lab, and sorry for the delayed reply. I had a quick look and noticed a few points that may need adjustment:

1. For the example notebook, I would suggest using the thyroid/aorta view, which is closer to the original dataset used in Speckle2Self. For other views, some fine-tuning may be needed to obtain better results.

2. The output may benefit from gamma correction; otherwise, the overall image can look too bright. This is also done in the inference code of the original repo.

3. The paper citation seems to be hallucinated, and it appears to cite another paper with the same title/name from the remote sensing field.

Other than these points, the integration looks great. Thanks again for your work!

> Hi @noseefood! Attempted to integrate Speckle2Self ([GitHub](https://github.com/noseefood/speckle2self)) into [zea](https://joss.theoj.org/papers/10.21105/joss.09881). See [example notebook](https://zea--377.org.readthedocs.build/en/377/notebooks/models/speckle2self_despeckling_example.html) on some of our data. Could you comment on whether the pre/post processing steps for the data are correctly implemented for this method? And is the output as expected or can we improve?

**Comment by tristan-deep** [comment]
> Hi, zea teams. Thanks a lot for integrating Speckle2Self into zea! I’m very happy to see this implementation. Last week I was not in the lab, and sorry for the delayed reply. I had a quick look and noticed a few points that may need adjustment:
> 
> 1. For the example notebook, I would suggest using the thyroid/aorta view, which is closer to the original dataset used in Speckle2Self. For other views, some fine-tuning may be needed to obtain better results.
> 2. The output may benefit from gamma correction; otherwise, the overall image can look too bright. This is also done in the inference code of the original repo.
> 3. The paper citation seems to be hallucinated, and it appears to cite another paper with the same title/name from the remote sensing field.
> 
> Other than these points, the integration looks great. Thanks again for your work!

Awesome @noseefood, thanks for taking a look!  
- As you suggested, I have updated the [example notebook](https://zea--358.org.readthedocs.build/en/358/notebooks/models/speckle2self_despeckling_example.html) with a different view serving as the example. That did seem to improve the output (also added more examples above in the PR description). Finetuning is also a good idea, will see if we can add a small script for that in the future. In the last cell we already apply gamma correction, let me know if you would apply it differently. 
- Sorry about the references, I copied those to hastily! I have checked now the citations in every spot and seem to be correct now ([notebook](https://zea--358.org.readthedocs.build/en/358/notebooks/models/speckle2self_despeckling_example.html), [class](https://zea--358.org.readthedocs.build/en/358/_autosummary/zea.models.speckle2self.html#zea.models.speckle2self.Speckle2Self), [module](https://zea--358.org.readthedocs.build/en/358/_autosummary/zea.models.speckle2self.html#module-zea.models.speckle2self)).

Note that these link to a version of the docs not merged to main yet!

I have one question remaining: on [your GitHub](https://github.com/noseefood/speckle2self) it states that: 
> _For optimal performance, please use ultrasound envelope data (not B-mode)_

Do you mean in this case log-compressed envelope data? That is currently how I have implemented it. This would be consistent from the sample image you use as well. I'm asking because for me log-compressed envelope data would constitute as B-mode data, but I know the definition of B-mode isn't very strict. Or do you mean rather to apply envelope data without log compression? It does also seem to kind of work but then I wasn't sure about the pre and post processing steps. The output of the network seems to be log-compressed already in any of the two cases.

**Comment by noseefood** [comment]
> > Hi, zea teams. Thanks a lot for integrating Speckle2Self into zea! I’m very happy to see this implementation. Last week I was not in the lab, and sorry for the delayed reply. I had a quick look and noticed a few points that may need adjustment:
> > 
> > 1. For the example notebook, I would suggest using the thyroid/aorta view, which is closer to the original dataset used in Speckle2Self. For other views, some fine-tuning may be needed to obtain better results.
> > 2. The output may benefit from gamma correction; otherwise, the overall image can look too bright. This is also done in the inference code of the original repo.
> > 3. The paper citation seems to be hallucinated, and it appears to cite another paper with the same title/name from the remote sensing field.
> > 
> > Other than these points, the integration looks great. Thanks again for your work!
> 
> Awesome @noseefood, thanks for taking a look!
> 
> * As you suggested, I have updated the [example notebook](https://zea--358.org.readthedocs.build/en/358/notebooks/models/speckle2self_despeckling_example.html) with a different view serving as the example. That did seem to improve the output (also added more examples above in the PR description). Finetuning is also a good idea, will see if we can add a small script for that in the future. In the last cell we already apply gamma correction, let me know if you would apply it differently.
> * Sorry about the references, I copied those to hastily! I have checked now the citations in every spot and seem to be correct now ([notebook](https://zea--358.org.readthedocs.build/en/358/notebooks/models/speckle2self_despeckling_example.html), [class](https://zea--358.org.readthedocs.build/en/358/_autosummary/zea.models.speckle2self.html#zea.models.speckle2self.Speckle2Self), [module](https://zea--358.org.readthedocs.build/en/358/_autosummary/zea.models.speckle2self.html#module-zea.models.speckle2self)).
> 
> Note that these link to a version of the docs not merged to main yet!
> 
> I have one question remaining: on [your GitHub](https://github.com/noseefood/speckle2self) it states that:
> 
> > _For optimal performance, please use ultrasound envelope data (not B-mode)_
> 
> Do you mean in this case log-compressed envelope data? That is currently how I have implemented it. This would be consistent from the sample image you use as well. I'm asking because for me log-compressed envelope data would constitute as B-mode data, but I know the definition of B-mode isn't very strict. Or do you mean rather to apply envelope data without log compression? It does also seem to kind of work but then I wasn't sure about the pre and post processing steps. The output of the network seems to be log-compressed already in any of the two cases.

Hi, it looks better now. Thank you again for your work!
Yes, by “ultrasound envelope data (not B-mode)” we mean log-compressed envelope data. More precisely, we refer to B-mode images after envelope detection and log compression, but without additional device-side image enhancement, except TGC. So your current implementation should be consistent with what we intended.
